astropy/table/_dataframes.py

```diff
@@ -0,0 +1,517 @@
+            dtype = array[name].dtype
+            if (dtype := array[name].dtype).kind not in INTEGER_DTYPE_KINDS:
+                continue
+
+            df_native[name] = df_native[name].astype(_numpy_to_pandas_dtype(dtype))
+
+    # Pandas-like index
+    if index:
+        df_native.set_index(index, inplace=True, drop=True)
+
+    return df_native
+
+
+def from_df(
+    df: Any, *, index: bool = False, units: Mapping[str, "UnitLike"] | None = None
+) -> "Table":
+    """Create a Table from any narwhals-compatible DataFrame."""
+    from .table import Table
+
+    if not HAS_NARWHALS:
+        raise ModuleNotFoundError(
+            "The narwhals library is required for the generic from_df method. "
+            "If you want to only convert from pandas, use the `from_pandas` method instead."
+        )
+    import narwhals as nw
+
+    # Create output
+    out = OrderedDict()
+
+    # Handle pandas index
+    if index:
+        if not hasattr(df, "index"):
+            raise ValueError(
+                "The input dataframe does not have an index. "
+                "Are you trying to convert a non-pandas dataframe? "
+                "Set `index=False` to avoid this error."
+            )
+        index_name = df.index.name or "index"
+        while index_name in df.columns:
+            index_name = "_" + index_name + "_"
+        df.reset_index(index_name, inplace=True, drop=False)
+
+    # Narwhals layer, must convert to eager
+    df_nw = nw.from_native(df)
+    if isinstance(df_nw, nw.LazyFrame):
+        df_nw = df_nw.collect()
+
+    # Handle units
+    if units is None:
+        units = {}
+    elif not isinstance(units, Mapping):
+        raise TypeError('Expected a Mapping "column-name" -> "unit"')
+    not_found = set(units.keys()) - set(df_nw.columns)
+    if not_found:
+        warnings.warn(f"`units` contains additional columns: {not_found}")
+
+    # Iterate over Narwhals columns
+    for column in df_nw.iter_columns():
+        # Unpack relevant data
+        data = column.to_numpy()
+        name = column.name
+        dtype = column.dtype
+        mask = column.is_null().to_numpy()
+        unit = units.get(name)
+
+        if isinstance(dtype, nw.Int128):
+            raise ValueError(
+                "Numpy and astropy Table do not support Int128, please use a smaller integer type."
+            )
+
+        # Handle nullable integers
+        if dtype.is_integer() and mask.any():
+            nullfilled = column.fill_null(0).to_numpy()
+            out[name] = MaskedColumn(data=nullfilled, mask=mask, unit=unit, copy=False)
+            continue
+
+        # Handle datetime columns
+        elif isinstance(dtype, nw.Datetime):
+            from astropy.time import Time
+
+            datetime = Time(data, format="datetime64")
+            datetime.format = "isot"
+            out[name] = datetime
+            continue
+
+        # Handle timedelta columns
+        elif isinstance(dtype, nw.Duration):
```

Comment:
Is this equivalent ? (I'm learning :))

```suggestion
        elif dtype.is_temporal():
```